Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,640 @@
+
+    static LineVerificationStatus[] RenderPerLineDiagnostics(
+      VerificationDiagnosticsParams verificationDiagnosticsParams,
+      VerificationTree[] verificationTrees,
+      int numberOfLines,
+      int numberOfResolutionErrors,
+      bool verificationStarted,
+      Container<Diagnostic> diagnostics) {
+      var result = new LineVerificationStatus[numberOfLines];
+
+      if (verificationTrees.Length == 0 && numberOfResolutionErrors == 0 && verificationStarted) {
+        for (var line = 0; line < numberOfLines; line++) {
+          result[line] = LineVerificationStatus.Verified;
+        }
+
+        return result;
+      }
+
+      // Render verification tree content into lines.
+      foreach (var verificationTree in verificationTrees) {
+        if (verificationTree.Filename == verificationDiagnosticsParams.Uri.GetFileSystemPath() ||
+            "untitled:" + verificationTree.Filename == verificationDiagnosticsParams.Uri) {
+          verificationTree.RenderInto(result);
+        }
+      }
+
+      // Fill in the missing "Unknown" based on the surrounding content
+      // The filling only takes Verified an Error
+      var previousNotUnknown = LineVerificationStatus.Unknown;
+      var lineDelta = 1;
+      // Two passes so that we can fill gaps based on what happened before AND after
+      for (var line = 0; 0 <= line; line += lineDelta) {
+        if (line == numberOfLines) {
+          lineDelta = -1;
+          previousNotUnknown = LineVerificationStatus.Unknown;
+          continue;
+        }
+        if (previousNotUnknown != LineVerificationStatus.Verified &&
+            previousNotUnknown != LineVerificationStatus.VerifiedObsolete &&
+            previousNotUnknown != LineVerificationStatus.VerifiedVerifying) {
+          previousNotUnknown = LineVerificationStatus.Unknown;
+        }
+        if (result[line] == LineVerificationStatus.Unknown) {
+          result[line] = previousNotUnknown;
+        } else {
+          previousNotUnknown = result[line];
+        }
+      }
+
+      var resolutionErrorRendered = 0;
+      foreach (var diagnostic in diagnostics) {
+        if (resolutionErrorRendered >= numberOfResolutionErrors) {
+          break;
+        }
+        if (diagnostic.Range.Start.Line >= 0 && diagnostic.Range.Start.Line < result.Length) {
+          result[diagnostic.Range.Start.Line] = LineVerificationStatus.ResolutionError;
+          resolutionErrorRendered++;
+        }
+      }
+
+      var existsErrorRange = false;
+      var existsError = false;
+      foreach (var line in result) {
+        existsErrorRange = existsErrorRange || line == LineVerificationStatus.ErrorContext;
+        existsError = existsError || line == LineVerificationStatus.AssertionFailed;
+      }
+
+      if (existsErrorRange && !existsError) {
+        existsError = false;
+      }
+
+      return result;
+    }
+  }
+
+
+  public enum VerificationStatus {
+    Unknown = 0,
```

Comment:
To the user, I would think both Unknown and Verified+Obsolete mean "Don't worry about it yet". Do you need both?

Reply:
I see what you mean. That's always the problem of null versus undefined.

Let me rename "Unknown" to "Nothing", because it's the initialization value for every line.